### ISSUE TITLE
Clarification on effect of time_distributed in keras

### ISSUE THREAD

**Issue by skeydan** [rstudio/keras#467]
Most blog posts on time series multi-step-ahead prediction with Keras (using R or Python) use `time_distributed`to connect the sequence output from the last RNN layer to the `dense` output (cf. the often cited 
https://machinelearningmastery.com/timedistributed-layer-for-long-short-term-memory-networks-in-python/
).

There is confusion about the purpose of using  `time_distributed(layer_dense(...))` instead of `layer_dense(...)` only. Evidently with Keras > 2.0, there is no difference between both versions (cf. 
https://stackoverflow.com/questions/47305618/what-is-the-role-of-timedistributed-layer-in-keras 
and e.g.  
https://github.com/keras-team/keras/issues/278
).

Just to be "empirically convinced", here are 2 tests on synthetic data that show that:
1) the structure of the models (parameters) is the same in both cases, and
2) loss is very nearly the same for both versions.

We separately test for the frequent use case where we're predicting a single series (example 1) and for predicting  two series at the same time (example 2). We're trying to reduce random effects by setting a seed and initializing the weights to 0 (something you would not want to do in real life).
We're looping over different sizes of the timesteps dimension and clearly see the above points confirmed.

Aside: The weight matrices themselves are not identical and may even be very different - just to achieve the same result - as we're still not eliminating all sources of random (`time_distributed` uses an RNN cell internally on whose initialization we can't exert any influence).

Example 1

```
library(keras)

n_tsteps <- 3:10
  
batch_size <- 10
n_epochs <- 200

data <- rep(1:7, 10) + rnorm(70) %>% round(digits = 2)

for (t in seq_along(n_tsteps)) {
  print(n_tsteps[t])
  
  data_matrix <-
    sapply(1:(length(data) - n_tsteps[t] * 2 + 1), function(index) {
      data[index:(index + n_tsteps[t] * 2 - 1)]
    }) %>% t()
  data_matrix
  x <- data_matrix[, 1:n_tsteps[t]]
  y <- data_matrix[, (n_tsteps[t] + 1):ncol(data_matrix)]
  dim(x) <- c(dim(x)[1], dim(x)[2], 1)
  dim(y) <- c(dim(y)[1], dim(y)[2], 1)
  
  
  set.seed(7777)
  model1 <- keras_model_sequential()
  model1 %>% time_distributed(layer_dense(units = 1, kernel_initializer = initializer_zeros(), input_shape = c(n_tsteps[t], 1)))
  model1 %>% compile(loss = "mse", optimizer = "adam")
  history <-
    model1 %>% fit(x,
                   y,
                   epochs = n_epochs,
                   batch_size = batch_size,
                   verbose = 0)
  plot(history)
  model1 %>% print()
  model1$get_weights()  %>% print()
  model1 %>% evaluate(x, y) %>% print()
  
  set.seed(7777)
  model2 <- keras_model_sequential()
  model2 %>% layer_dense(units = 1, input_shape = c(n_tsteps[t], 1), kernel_initializer = initializer_zeros())
  model2 %>% print()
  model2 %>% compile(loss = "mse", optimizer = "adam")
  history <-
    model2 %>% fit(x,
                   y,
                   epochs = n_epochs,
                   batch_size = batch_size,
                   verbose = 0)
  plot(history)
  model2$get_weights()  %>%  print()
  model2 %>% evaluate(x, y) %>% print()
  
}
```
Example 2:

```
library(keras)

n_tsteps <- 3:10

n_features <- 2
batch_size <- 10
n_epochs <- 200

col1 <- rep(1:7, 10) + rnorm(70) %>% round(digits = 2)
col2 <- rep(1:70) + rnorm(70) %>% round(digits = 2)

for (t in seq_along(n_tsteps)) {
  print(n_tsteps[t])
  
  col1_matrix <-
    sapply(1:(length(col1) - n_tsteps[t] * 2 + 1), function(index) {
      col1[index:(index + n_tsteps[t] * 2 - 1)]
    }) %>% t()
  col1_matrix
  
  col2_matrix <-
    sapply(1:(length(col2) - n_tsteps[t] * 2 + 1), function(index) {
      col2[index:(index + n_tsteps[t] * 2 - 1)]
    }) %>% t()
  col2_matrix
  
  data_array <-
    array(0, dim = c(nrow(col1_matrix), n_tsteps[t] * 2, n_features))
  data_array[, , 1] <- col1_matrix
  data_array[, , 2] <- col2_matrix
  
  x <- data_array[, 1:n_tsteps[t],]
  y <- data_array[, (n_tsteps[t] + 1):ncol(data_array),]
  
  set.seed(7777)
  model1 <- keras_model_sequential()
  model1 %>% time_distributed(layer_dense(units = n_features,
                                          kernel_initializer = initializer_zeros()),
                              input_shape = c(n_tsteps[t], n_features))
  model1 %>% print()
  model1 %>% compile(loss = "mse", optimizer = "adam")
  history <-
    model1 %>% fit(x,
                   y,
                   epochs = n_epochs,
                   batch_size = batch_size,
                   verbose = 0)
  plot(history)
  model1$get_weights()  %>%  print()
  model1 %>% evaluate(x, y) %>% print()
 
  set.seed(7777)
  model2 <- keras_model_sequential()
  model2 %>% layer_dense(units = n_features,
                         input_shape = c(n_tsteps[t], n_features),
                         kernel_initializer = initializer_zeros())
  model2 %>% print()
  model2 %>% compile(loss = "mse", optimizer = "adam")
  history <-
    model2 %>% fit(x,
                   y,
                   epochs = n_epochs,
                   batch_size = batch_size,
                   verbose = 0)
  plot(history)
  model2$get_weights()  %>%  print()
  model2 %>% evaluate(x, y) %>% print()
}
```
